sbnobj/Common/CRT/CRTHit.hh

```diff
@@ -26,12 +26,12 @@ namespace sbn::crt {
       std::map< uint8_t, std::vector<std::pair<int,float> > > pesmap;
       float peshit;
      
-      uint32_t ts0_s;
+      uint64_t ts0_s;
       int8_t ts0_s_corr;
       
-      uint32_t ts0_ns;
-      int32_t ts0_ns_corr;
-      int32_t ts1_ns;
+      uint64_t ts0_ns;
+      int64_t ts0_ns_corr;
+      int64_t ts1_ns;
```

Comment:
Could you explain the need for this change?
There is no inline documentation, so I can only guess from the data member names: the `_ns` ones represent the number of nanoseconds of a timestamp since the last round second. So these are in the `[ 0 ; 1'000'000'000 [` range.
I _suppose_ that the `_s` part is the number of seconds since some Epoch, so unless you pick as Epoch a day in the 15th century, 32 bit should suffice (they span 68 years if signed, twice that much if unsigned). Ans in no case the nanosecond part requires 64 bit, unless it is _the full timestamp_ expressed in nanoseconds.
If 64-bit timestamps are desired, I'd rather recommend two member functions:
```cpp
      int64_t ts0() const { return static_const<int64_t>(ts0_s) * 1'000'000'000LL + ts0_ns; }
      int64_t ts1() const { return static_const<int64_t>(ts0_s) * 1'000'000'000LL + ts1_ns; }
```
(I did not apply the "correction" `ts0_s_corr` because I don't know if one should here).

Reply:
Answer to 1 : T0 is an absolute timestamp, i.e. number of ns since beginning of 1970. A big number. It is also corrected for the fact that PPS takes time to arrive to various FEBs. In CRTData we have assigned T0 to uint64_t, however we pass this to CRTHit it was assigned to uint32_t, So there is a need of change in the CRTHit. I am not sure the use of `ts0_s_corr` I am trying to figure it out, and revisiting the CRTHitRecoalg, will have more information in day or two and love to discuss if there are better ways to improve. Thanks a lot for helping/guiding me.

Reply:
If `ts0_ns` are the full timestamps, then there should be no need for `ts0_s`, which is just `ts0_ns / 1'000'000'000`.
And anyway `ts0_s` should not need 64 bit if it's only the number of seconds.
Can you show me a producer module where this data is filled?

Reply:
Here is the Algorithm accessing these variables.
https://github.com/SBNSoftware/icaruscode/blob/develop/icaruscode/CRT/CRTUtils/CRTHitRecoAlg.cc#L255

Yes, you are correct, as we have ts0_ns, it can be extracted.  As it was there for long time back and used by both detector, we have not changed this variable. It will be better to be removed.

Reply:
Thank you for the pointer to the code, it is much clearer now.
The primary source of my confusion is that in other data products when there are two fields like `t0_s` and `t0_ns`, the first contains only the seconds of the time stamp from the Epoch, and the second only the nanoseconds after that second, so the full timestamp is `t0_s * 1'000'000'000ULL + t0_ns` (and note that I need to specify that the 10⁹ factor is a 64-bit number, unsigned long long, with `ULL` suffix).
If you have reasons to nevertheless keep both the absolute timestamp (`t0_ns`) _and_ the second-only part (`t0_s`), so ahead with it. But please add inline documentation explaining all of these conventions.
For example:
```cpp
      uint64_t ts0_s;  ///< Second-only part of timestamp T0.
      int8_t ts0_s_corr; ///< [what is this??]

      uint64_t ts0_ns;  ///< Timestamp T0 (from White Rabbit), in UTC absolute time scale in nanoseconds from the Epoch.
      int64_t ts0_ns_corr; ///< [what is this??]
      int64_t ts1_ns; ///< Timestamp T1 ([?]), in UTC absolute time scale in nanoseconds from the Epoch.
```
Also note that the code you pointed out has a minor issue: the conversions in seconds unnecessarily convert the number in `double` and then back to `int` because `1e-9` is a double (should go with `/ 1'000'000'000` instead); this is probably not fatal because there should still be no loss of precision _in the second part_ of the timestamp.